### ISSUE TITLE
[REQUEST]: language detection

### ISSUE THREAD

**Issue by verhovsky** [compiler-explorer/compiler-explorer#6687]
### Is your feature request related to a problem? Please describe

I would like it if I didn't have to click the language drop down and scroll through all the languages when I paste some code.

### Describe the solution you'd like

The language of the code window should be detected automatically any time I replace all the code with pasting or clear it. For languages that are supersets like C and C++ it should still get narrowed down in the drop down because even in a line or two of C code you can clearly tell it's definitely not Python.

### Describe alternatives you've considered

Not applicable

### Additional context

Not applicable

**Comment by mattgodbolt** [comment]
What if the "guess" is wrong? Wouldn't that be more annoying? How might we tell the difference from C. C++ and the other C-like languages where it's somewhat ambiguous? I very often want to compile plain code as C++ even if it doesn't use a C++ construct (like `template`, `class` etc).

**Comment by verhovsky** [comment]
If the detected language is wrong then you would just have to do what you have to do 100% of the time now. assuming detection doesn't add any overhead, it's a strictly better situation with less clicks.

Most programmers use language detection every day when creating a new file in VS Code, it's right almost all the time. There's a couple cases of supersets: c/c++, js/ts and groovy/java (not on compiler explorer) https://guesslang.readthedocs.io/en/latest/ . These are a small number of ambiguous pairs, it's not going to mis-guess C as Haskell. But yes C as C++ will be a common issue, you could make it that when C or C++ is detected the drop down would put all the other C-like languages at the top of the drop down and grey out the rest of the values in the dropdown and if the buffer is set to C then pasting some new code that's detected as ambiguously either C or C++ (the library provides probability estimates) would just keep whatever language it already was.

Basically, if you're concerned about language detection being wrong, then you have to only change the language when the language detection is sure it's not wrong.
